perf: improve ManagedAuthenticatedEncryptor Decrypt() and Encrypt() flow

DeagleGross opened the pull request:
The goal of PR is to bring linux performance closer to windows performance for `DataProtection` scenario. Below is the picture of [Antiforgery](https://github.com/aspnet/Benchmarks/blob/main/src/BenchmarksApps/Antiforgery/Program.cs) benchmarks on win vs lin machines.
![image](https://github.com/user-attachments/assets/9c1240cb-86f2-4ae7-8281-62930fe5bbeb)

## Results: DataProtection Benchmark
The benchmark I am relying on to show the result numbers is [here](https://github.com/DeagleGross/Baraholka/blob/main/Benchmarks/Benchmarks/DataProtectionBenchmarks.cs), which is basically building default `ServiceProvider`, adding DataProtection via `.AddDataProtection()` and calling `IDataProtector.Protect()` or `IDataProtector.Unprotect()`. 

_example crank run:_
> crank --config https://raw.githubusercontent.com/DeagleGross/Baraholka/refs/heads/main/Benchmarks/Benchmarks/crank/data-protector.yml --scenario dataprotector --profile aspnet-perf-lin --application.framework net10.0 --application.runtimeVersion "10.0.0-preview.3.25156.13" --application.options.outputFiles "D:\code\aspnetcore\artifacts\bin\Microsoft.AspNetCore.DataProtection\Release\net10.0\\*.dll" --application.options.outputFiles "D:\code\aspnetcore\artifacts\bin\Microsoft.AspNetCore.DataProtection.Abstractions\Release\net10.0\\*.dll" --application.variables.filterArg "*DataProtectionBenchmarks*"

| net | Code | Method    | Mean      | Error     | StdDev    | Gen0   | Allocated | Diff (% µs) |
| --- | ------ | ---------- |----------:|----------:|----------:|-------:|----------:|------:|
| net10 | main branch |  Unprotect |  9.706 µs | 0.0143 µs | 0.0134 µs | 0.1068 |   3.34 KB | - |
| net10 | changed |  Unprotect | 7.033 µs | 0.0104 µs | 0.0098 µs | 0.2518 |   1.57 KB | 38.5% |
| net10 | main branch |  Protect | 10.88 µs | 0.015 µs | 0.014 µs | 0.6256 |   3.91 KB | - | 
| net10 | changed | Protect   | 7.888 µs | 0.0141 µs | 0.0132 µs | 0.3357 |   2.15 KB | 37.93% |

## Results: Antiforgery Benchmark

However, since we are originally looking at improving the Antiforgery performance on linux, I ran the [Antiforgery benchmark](https://github.com/aspnet/Benchmarks/blob/main/src/BenchmarksApps/Antiforgery/Program.cs) including locally built dll's from this PR. 

_example crank run:_
  > crank --config https://raw.githubusercontent.com/aspnet/Benchmarks/refs/heads/main/scenarios/antiforgery.benchmarks.yml --scenario antiforgery-validation --profile aspnet-perf-lin --application.framework net10.0 --application.runtimeVersion "10.0.0-preview.3.25156.13" --application.options.outputFiles "D:\code\aspnetcore\artifacts\bin\Microsoft.AspNetCore.DataProtection\Release\net10.0\\*.dll" --application.options.outputFiles "D:\code\aspnetcore\artifacts\bin\Microsoft.AspNetCore.DataProtection.Abstractions\Release\net10.0\\*.dll"

current aspnet core gives these stats on the benchmark (avg on 4 runs):
| Stat | Value | 
| --- | --- |
| Requests/sec            | 111,771 |
| Max Working Set (MB)    | 141.5 |
| Max Private Memory (MB) | 491 |

RPS of run with changed dll's varies from run to run, therefore I ran it 10 times
| # | Stat | Value |
| --- | --- | --- | 
| 1 | Requests/sec            | 163,286             |
| 2 | Requests/sec            | 163,648             |
| 3 | Requests/sec            | 164,287             |
| 4 | Requests/sec            | 162,611             |
| 5 | Requests/sec            | 163,820             |
| 6 | Requests/sec            | 164,098             |
| 7 | Requests/sec            | 163,480             |
| 8 | Requests/sec            | 162,668             |
| 9 | Requests/sec            | 163,302             |
| 10 | Requests/sec          | 161,953             |
| **Mean:** 163,215.3  | **Mean Diff %:** 46.06%  |

But memory usage is stable with such values:
| Stat | Value | diff |
| --- | --- | --- |
| Max Working Set (MB)    | 127 | 9.3% |
| Max Private Memory (MB) | 442 | 10% |

**Which provide an evidence of ~10% of max app allocation size and ~46% RPS improvement.**

## Optimization details
Another improvement can be achieved after new APIs introduced in dotnet/runtime: https://github.com/dotnet/runtime/issues/111154

I looked into `Unprotect` method for `ManagedAuthenticatedEncryptor` and spotted `MemoryStream` usage and multiple `Buffer.BlockCopy` usages. Also I saw that there is some shuffling of `byte[]` data, which I think can be skipped and performed in such a way, that some allocations are skipped.

In order to be as safe as possible, I created a separate `DataProtectionPool` which provides API to rent and return byte arrays. It is not intersecting with `ArrayPool<byte>.Shared`.

1) `ManagedSP800_108_CTR_HMACSHA512.DeriveKeys` is changed to explicit usage `ManagedSP800_108_CTR_HMACSHA512.DeriveKeysHMACSHA512`, because [_kdkPrfFactory](https://source.dot.net/#Microsoft.AspNetCore.DataProtection/Managed/ManagedAuthenticatedEncryptor.cs,29) is anyway hardcoded to use `HMACSHA512`. There is a static API allowing to hash without allocating kdk `byte[]` which is rented from the buffer: `HMACSHA512.TryHashData(kdk, prfInput, prfOutput, out _);`

2) Avoided usage of `DeriveKeysWithContextHeader` which allocates a separate intermediate array for `contextHeader` and `context`. Instead passing the spans `operationSubkey` and `validationSubkey` directly into `ManagedSP800_108_CTR_HMACSHA512.DeriveKeys`

3) `ManagedSP800_108_CTR_HMACSHA512.DeriveKeysHMACSHA512` had 2 more arrays (`prfInput` and `prfOutput`), which now I am renting (via `DataProtectionPool`) or even `stackalloc`'ing. They are returned to the pool with `clearArray: true` flag to make sure key material is removed from the memory after usage.

4) In `Decrypt()` flow I am again using [HashAlgorithm.TryComputeHash](https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.hashalgorithm.trycomputehash?view=net-9.0) overload, which works based on the `Span<byte>` types, compared to previously used [HashAlgorithm.ComputeHash](https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.hashalgorithm.computehash?view=net-9.0)

5) In `Decrypt()` flow changed usage to [SymmetricAlgorithm.DecryptCbc()](https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.symmetricalgorithm.decryptcbc?view=net-9.0#system-security-cryptography-symmetricalgorithm-decryptcbc(system-readonlyspan((system-byte))-system-readonlyspan((system-byte))-system-span((system-byte))-system-security-cryptography-paddingmode)) instead of [CryptoTransform.TransformBlock()](https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.icryptotransform.transformblock?view=net-9.0) with same idea to use `Span<byte>` API instead of another `byte[]` allocation.

6) `Encrypt()` flow is reusing №1, №2 and №3 optimizations as well

7) `Encrypt()` before was relying on the `MemoryStream` and `CryptoStream` to write data in the result buffer, but I am pre-calculating the length, and then doing a single allocation of result array: `var outputArray = new byte[keyModifierLength + ivLength + cipherTextLength + macLength];` All required data is copied into the outputArray via APIs supporting `Span<byte>`. 

All listed optimizations are included in the `net10` TFM, but only some (№ 2, №3 and №6) are used in `netstandard2.0` and `netFx` TFMs which DataProtection also targets.

Related to #59287

DeagleGross committed:
a bit of improvement

DeagleGross committed:
another buffer.blockCopy removal

DeagleGross committed:
use `DecryptCbc` instead?

DeagleGross committed:
tests and remove another array

DeagleGross committed:
fix the slicing of IV

DeagleGross committed:
slice correctly!

mgravell commented on src/DataProtection/DataProtection/src/Managed/ManagedAuthenticatedEncryptor.cs:
```diff
@@ -161,6 +166,8 @@ public byte[] Decrypt(ArraySegment<byte> protectedPayload, ArraySegment<byte> ad
         protectedPayload.Validate();
         additionalAuthenticatedData.Validate();
 
+        var protectedPayloadSpan = protectedPayload.AsSpan();
```
unused?

DeagleGross replied:
removed

mgravell commented on src/DataProtection/DataProtection/src/KeyManagement/KeyManagementOptions.cs:
```diff
@@ -111,7 +111,7 @@ internal static TimeSpan MaxServerClockSkew
     /// Settable for testing.
     /// </remarks>
-    internal TimeSpan DefaultKeyResolverRetryDelay { get; set; } = TimeSpan.FromMilliseconds(200);
+    internal TimeSpan DefaultKeyResolverRetryDelay { get; set; } = TimeSpan.FromMilliseconds(200.0);
```
nit: an odd thing to include in a PR; was this to make a warning happy?

DeagleGross replied:
yep, rollbacked it

mgravell commented on src/DataProtection/DataProtection/src/Managed/ManagedAuthenticatedEncryptor.cs:
```diff
@@ -233,27 +237,33 @@ public byte[] Decrypt(ArraySegment<byte> protectedPayload, ArraySegment<byte> ad
+                    return symmetricAlgorithm.DecryptCbc(ciphertext, iv);
+#else
+                    var iv = new byte[_symmetricAlgorithmBlockSizeInBytes];
+                    Buffer.BlockCopy(protectedPayload.Array!, ivOffset, iv, 0, iv.Length);
```
maybe still use span copy here?

mgravell commented on src/DataProtection/DataProtection/src/Managed/ManagedAuthenticatedEncryptor.cs:
```diff
@@ -187,8 +193,6 @@ public byte[] Decrypt(ArraySegment<byte> protectedPayload, ArraySegment<byte> ad
             }
 
             ArraySegment<byte> keyModifier = new ArraySegment<byte>(protectedPayload.Array!, keyModifierOffset, ivOffset - keyModifierOffset);
-            var iv = new byte[_symmetricAlgorithmBlockSizeInBytes];
```
just want to check: this is the main reduction (on suitable TFMs), yes?

DeagleGross replied:
there are a couple of `byte[]` allocations I am removing. I am going to explain it in details in the PR description. I will resolve comment once I have written the PR description

mgravell commented on src/DataProtection/DataProtection/src/Managed/ManagedAuthenticatedEncryptor.cs:
```diff
@@ -233,27 +237,33 @@ public byte[] Decrypt(ArraySegment<byte> protectedPayload, ArraySegment<byte> ad
+                    {
+                        var length = macOffset - ciphertextOffset;
+                        var result = new byte[length];
+                        _ = cryptoTransform.TransformBlock(protectedPayload.Array!, ciphertextOffset, length, result, 0);
```
should this be `TransformFinalBlock`? or a combination of both? I genuinely don't know, but: this needs to be 100% right. Perhaps check *exactly* what `Write` and `FlushFinalBlock` would do here.

mgravell commented on src/DataProtection/DataProtection/src/Managed/ManagedAuthenticatedEncryptor.cs:
```diff
@@ -233,27 +237,33 @@ public byte[] Decrypt(ArraySegment<byte> protectedPayload, ArraySegment<byte> ad
-                            cryptoStream.Write(protectedPayload.Array!, ciphertextOffset, macOffset - ciphertextOffset);
-                            cryptoStream.FlushFinalBlock();
+                    // note: here protectedPayload.Array is taken without an offset (cant use AsSpan() on ArraySegment)
+                    var ciphertext = protectedPayload.Array.AsSpan().Slice(ciphertextOffset, macOffset - ciphertextOffset);
```
move range to `AsSpan()` ?

DeagleGross replied:
not familiar with API, thanks for spotting, I changed to this overload

mgravell commented on src/DataProtection/DataProtection/src/Managed/ManagedAuthenticatedEncryptor.cs:
```diff
@@ -233,27 +237,33 @@ public byte[] Decrypt(ArraySegment<byte> protectedPayload, ArraySegment<byte> ad
-                            // At this point, outputStream := { plaintext }, and we're done!
-                            return outputStream.ToArray();
-                        }
+                    return symmetricAlgorithm.DecryptCbc(ciphertext, iv);
```
are we using CBC mode? I genuinely don't know - mostly after validation / explanation of why this choice when migrating - if this is what `CryptoStream` would use, maybe link?

Also, note that since this is a private method, we might also wonder whether we can remove the decrypted buffer by using the overload that takes a `Span<byte>` input, and moving the allocation (maybe as a lease) to the consuming code site; if the input is an oversized array, then note that `MemoryStream` allows an array to be passed *in* via the .ctor, which might mean it works fine in the fallback mode too, although that is usually intended for read buffers - just make sure you use the `writable: true` param and check the length written.

I can't tell you how the size of cipher-bytes and plain-bytes relate; is it 1:1? :shrug: don't know!

but, presumably, removing this result array would have a significant impact on the remaining allocs

DeagleGross replied:
thanks for suggestion: I used a new API using Spans and removing yet another allocation!

mgravell commented on src/DataProtection/DataProtection/src/Managed/ManagedAuthenticatedEncryptor.cs:
```diff
@@ -233,27 +237,33 @@ public byte[] Decrypt(ArraySegment<byte> protectedPayload, ArraySegment<byte> ad
                     }
 
                     // Step 5: Decipher the ciphertext and return it to the caller.
+#if NET10_0_OR_GREATER
```
TFM choice: do we support down-level? if not, just burn the old code; if we do: note that many of these APIs exist from net6 onwards, IIRC.

DeagleGross replied:
the code targets netFx, netstandard and **current** core:
`<TargetFrameworks>$(DefaultNetFxTargetFramework);netstandard2.0;$(DefaultNetCoreTargetFramework)</TargetFrameworks>`

I think there is no point in specifying lower TFMs because this code will not be backported (most probably). What do you think?

DeagleGross replied:
discussed offline: no value in bringing the other netCore TFMs in compatibility with this code

mgravell commented on src/DataProtection/DataProtection/src/Managed/ManagedAuthenticatedEncryptor.cs:
```diff
@@ -206,16 +210,16 @@ public byte[] Decrypt(ArraySegment<byte> protectedPayload, ArraySegment<byte> ad
                 try
                 {
                     _keyDerivationKey.WriteSecretIntoBuffer(new ArraySegment<byte>(decryptedKdk));
-                    ManagedSP800_108_CTR_HMACSHA512.DeriveKeysWithContextHeader(
```
any clues on the difference between these two?

mgravell commented on src/DataProtection/DataProtection/src/SP800_108/ManagedSP800_108_CTR_HMACSHA512.cs:
```diff
@@ -55,6 +55,55 @@ public static void DeriveKeys(byte[] kdk, ArraySegment<byte> label, ArraySegment
+                prfInput[3] = (byte)(i);
+
+                // Run the PRF and copy the results to the output buffer
+                var prfOutput = prf.ComputeHash(prfInput);
```
Is this `HashAlgorithm` (or sublass) ? If so, note that there is a `ComputeHash(byte[], int, int)` from .NET 9 onwards which would allo `prfInput` to be a pooled+oversized array, constrained i.e.

```
// See SP800-108, Sec. 5.1 for the format of the input to the PRF routine.
var prfLength = checked(sizeof(uint) /* [i]_2 */ + label.Length + 1 /* 0x00 */ + contextSharedLength + sizeof(uint) /* [K]_2 */)
#if NET9_0_OR_GREATER // check exact token
var prfInput = ArrayPool<byte>.Shared.Rent(prfLength);
#else
var prfInput = new byte[prfLength];
#endif

// ...

#if NET9_0_OR_GREATER
var prfOutput = prf.ComputeHash(prfInput, 0, prfLength);
ArrayPool<byte>.Shared.Return(prfInput, clearArray: true); // clearing because security-related
#else
var prfOutput = prf.ComputeHash(prfInput);
#endif
```

mgravell replied:
ooh, there's also `TryComputeHash` that allows a buffer to be passed in

DeagleGross replied:
that's a great suggestion, and I tried it, but it simply throws. There are a couple of allocations that should probably be rented via the ArrayPool, and it would bring the allocations down, but if I replace the code to rent and return, then it fails.

I am going to write down the findings and the problems here in a while

mgravell replied:
Yep; that's why when I use `ArrayPool<T>.Shared.Rent` I almost always call the local `oversized` - then I can't forget! For example: https://github.com/dotnet/extensions/blob/c08790cf79d2d8feaa53751f039f38638741a154/src/Libraries/Microsoft.Extensions.Caching.Hybrid/Internal/InbuiltTypeSerializer.cs#L43

There's also cheeky things you can do such as:

``` c#
byte[]? lease = null;
Span<byte> buffer = length < 128
    ? stackalloc byte[length]
    : (lease = ArrayPool<byte>.Shared.Rent(length)).AsSpan(0, length);

// code here using "buffer", which is right-sized and could be stack or heap

if (lease is not null)
{
    ArrayPool<byte>.Shared.Return(lease);
}
```

DeagleGross replied:
thanks for pointing it out, made an optimization

DeagleGross commented on src/DataProtection/DataProtection/src/Managed/ManagedAuthenticatedEncryptor.cs:
```diff
@@ -206,16 +210,16 @@ public byte[] Decrypt(ArraySegment<byte> protectedPayload, ArraySegment<byte> ad
                         prfFactory: _kdkPrfFactory,
                         output: new ArraySegment<byte>(derivedKeysBuffer));
 
-                    Buffer.BlockCopy(derivedKeysBuffer, 0, decryptionSubkey, 0, decryptionSubkey.Length);
```
Here is the benchmark I used to check the copy approaches - https://github.com/DeagleGross/Baraholka/blob/main/Benchmarks/Benchmarks/CopyBytesBenchmark.cs

        | Method           | N    | Mean      | Error     | StdDev    | Allocated |
        |----------------- |----- |----------:|----------:|----------:|----------:|
        | Array_Copy       | 10   |  2.402 ns | 0.0181 ns | 0.0142 ns |         - |
        | Buffer_BlockCopy | 10   |  3.475 ns | 0.0232 ns | 0.0217 ns |         - |
        | Span_CopyTo      | 10   |  1.126 ns | 0.0139 ns | 0.0124 ns |         - |
        | Array_Copy       | 100  |  3.755 ns | 0.0313 ns | 0.0277 ns |         - |
        | Buffer_BlockCopy | 100  |  4.188 ns | 0.0419 ns | 0.0392 ns |         - |
        | Span_CopyTo      | 100  |  2.660 ns | 0.0196 ns | 0.0183 ns |         - |
        | Array_Copy       | 1000 | 10.202 ns | 0.1154 ns | 0.0964 ns |         - |
        | Buffer_BlockCopy | 1000 | 12.023 ns | 0.1411 ns | 0.1251 ns |         - |
        | Span_CopyTo      | 1000 |  8.986 ns | 0.1847 ns | 0.1814 ns |         - |

DeagleGross committed:
try with encrypt

DeagleGross committed:
finish encrypt

DeagleGross committed:
use same overload in decrypt()

DeagleGross committed:
dont allocate for prf-output as well

DeagleGross committed:
address PR comments

DeagleGross committed:
remove spaces

DeagleGross committed:
prfInput and prfOutput rented

DeagleGross committed:
correctHash as rented

DeagleGross committed:
more details

DeagleGross committed:
use static `TryHashData` for specific hash implementation

DeagleGross committed:
prettify

GrabYourPitchforks commented:
### Code review notes

1. It's generally not advisable to pool buffers for sensitive cryptographic operations, such as those which perform key storage or manipulation. This tends to increase the attack surface of the application and should only be performed if it's absolutely required to meet some performance goal. If you absolutely must use pooled buffers, use different pooled buffers for key material specifically vs. (all other data). In dataprotection, key material would be the KDK, KEK, and individual decryption + validation keys.

2. This PR introduces a call to the `SymmetricKey.Key` property setter. The original code intentionally avoided calling this property setter because it duplicates the sensitive key material in such a way that the caller has no control over the new lifetime, and this undermines other protections present in the system (the use of the `Secret` type and the widespread use of pinning within the core crypto logic). Of course, the EG can always say that the new behavior is preferred over the old behavior, but there needs to be an _explicit acknolwedgement_ that there is a security tradeoff here.

It's possible we can make changes to the underlying `SymmetricAlgorithm` type to improve the perf without reducing the security stance, but this would require new API to be exposed within corelib. Since you're targeting these changes for net10 that's probably acceptable? Work with Jeff's team to give them your requirements and they can add the work to the backlog.

### Scenario notes

Is the goal to improve the performance of the \[real-world?\] AntiForgery benchmark or to improve the performance of DataProtection in a standalone benchmark? The PR description (and attached graph) make it sound like improving the performance of the crank-based benchmark is the goal, but no throughput measurement is provided for the changes in this PR. Please provide that graph. It would supply evidence that these changes have real-world impact and aren't just microbenchmark improvements.

DeagleGross commented:
> ### Code review notes
> 1. It's generally not advisable to pool buffers for sensitive cryptographic operations, such as those which perform key storage or manipulation. This tends to increase the attack surface of the application and should only be performed if it's absolutely required to meet some performance goal. If you absolutely must use pooled buffers, use different pooled buffers for key material specifically vs. (all other data). In dataprotection, key material would be the KDK, KEK, and individual decryption + validation keys.
> 2. This PR introduces a call to the `SymmetricAlgorithm.Key` property setter. The original code intentionally avoided calling this property setter because it duplicates the sensitive key material in such a way that the caller has no control over the new lifetime, and this undermines other protections present in the system (the use of the `Secret` type and the widespread use of pinning within the core crypto logic). Of course, the EG can always say that the new behavior is preferred over the old behavior, but there needs to be an explicit acknowledgement that there is a security tradeoff here. The tradeoff shouldn't be a mere side effect that is likely to go unnoticed.
>    It's possible we can make changes to the underlying `SymmetricAlgorithm` type to improve the perf without reducing the security stance, but this would require new API to be exposed within corelib. Since you're targeting these changes for net10 that's probably acceptable? Work with Jeff's team to give them your requirements and they can add the work to the backlog.
> 3. The pattern `if (len < CONST) { foo = stackalloc[len]; } else { foo = Rent(len).Slice(len); }` is typically considered an antipattern. Prefer a pattern like `if (len < CONST) { foo = stackalloc[CONST]; } else { foo = Rent(len); } foo = foo.Slice(len);` instead. (Basically, the stackalloc should be a const, not variable-length.)
> 
> ### Scenario notes
> Is the goal to improve the performance of the [real-world?] AntiForgery benchmark or to improve the performance of DataProtection in a standalone benchmark? The PR description (and attached graph) make it sound like improving the performance of the crank-based benchmark is the goal, but no throughput measurement is provided for the changes in this PR. Please provide that graph. It would supply evidence that these changes have real-world impact and aren't just microbenchmark improvements.

Thanks for detailed answer @GrabYourPitchforks! Firstly, I ran the **Antiforgery** benchmark multiple times, and I provided the results in the PR description. 

Re №3: I ran a [BenchmarkDotNet](https://github.com/DeagleGross/Baraholka/blob/main/Benchmarks/Benchmarks/StackAllocCost.cs) for stackalloc with dynamic \ constant length of stackalloc (also with or without `[SkipLocalsInit]`): results are that without `[SkipLocalsInit]` dynamic length is even faster, but with `[SkipLocalsInit]` you are correct. I will probably enable the `[SkipLocalsInit]` on the method and then we can use constant length. What do you think?

Re №2: Thanks for clarifying it, I will create issues on the dotnet/runtime explaining what API I would like to have to make DataProtection's flow dont use `byte[]` directly.

Re №1: Could you please describe how attack surface of the application is increased if pool buffers are used? Does that mean that pooling is easier to inject into via reflection for example? Actually, even if we will not be using pooling byte arrays, if I work with corelib to introduce APIs supporting `Span<byte>`, we can choose to `stackalloc` or allocate a new byte via `new byte[]`, meaning that we will not ever touch pool buffers. Does it make sense and is that more secure? I think most of the code is not using "big" arrays (say with length more than 128), and we will hit stackalloc in most cases.
```
Span<byte> arr = length <= 128
    ? stackalloc byte[128]
    : new byte[length];
```

DeagleGross commented:
> > ### Code review notes
> > 
> > 1. It's generally not advisable to pool buffers for sensitive cryptographic operations, such as those which perform key storage or manipulation. This tends to increase the attack surface of the application and should only be performed if it's absolutely required to meet some performance goal. If you absolutely must use pooled buffers, use different pooled buffers for key material specifically vs. (all other data). In dataprotection, key material would be the KDK, KEK, and individual decryption + validation keys.
> > 2. This PR introduces a call to the `SymmetricAlgorithm.Key` property setter. The original code intentionally avoided calling this property setter because it duplicates the sensitive key material in such a way that the caller has no control over the new lifetime, and this undermines other protections present in the system (the use of the `Secret` type and the widespread use of pinning within the core crypto logic). Of course, the EG can always say that the new behavior is preferred over the old behavior, but there needs to be an explicit acknowledgement that there is a security tradeoff here. The tradeoff shouldn't be a mere side effect that is likely to go unnoticed.
> >    It's possible we can make changes to the underlying `SymmetricAlgorithm` type to improve the perf without reducing the security stance, but this would require new API to be exposed within corelib. Since you're targeting these changes for net10 that's probably acceptable? Work with Jeff's team to give them your requirements and they can add the work to the backlog.
> > 3. The pattern `if (len < CONST) { foo = stackalloc[len]; } else { foo = Rent(len).Slice(len); }` is typically considered an antipattern. Prefer a pattern like `if (len < CONST) { foo = stackalloc[CONST]; } else { foo = Rent(len); } foo = foo.Slice(len);` instead. (Basically, the stackalloc should be a const, not variable-length.)
> > 
> > ### Scenario notes
> > Is the goal to improve the performance of the [real-world?] AntiForgery benchmark or to improve the performance of DataProtection in a standalone benchmark? The PR description (and attached graph) make it sound like improving the performance of the crank-based benchmark is the goal, but no throughput measurement is provided for the changes in this PR. Please provide that graph. It would supply evidence that these changes have real-world impact and aren't just microbenchmark improvements.
> 
> Thanks for detailed answer @GrabYourPitchforks! Firstly, I ran the **Antiforgery** benchmark multiple times, and I provided the results in the PR description.
> 
> Re №3: I ran a [BenchmarkDotNet](https://github.com/DeagleGross/Baraholka/blob/main/Benchmarks/Benchmarks/StackAllocCost.cs) for stackalloc with dynamic \ constant length of stackalloc (also with or without `[SkipLocalsInit]`): results are that without `[SkipLocalsInit]` dynamic length is even faster, but with `[SkipLocalsInit]` you are correct. I will probably enable the `[SkipLocalsInit]` on the method and then we can use constant length. What do you think?
> 
> Re №2: Thanks for clarifying it, I will create issues on the dotnet/runtime explaining what API I would like to have to make DataProtection's flow dont use `byte[]` directly.
> 
> Re №1: Could you please describe how attack surface of the application is increased if pool buffers are used? Does that mean that pooling is easier to inject into via reflection for example? Actually, even if we will not be using pooling byte arrays, if I work with corelib to introduce APIs supporting `Span<byte>`, we can choose to `stackalloc` or allocate a new byte via `new byte[]`, meaning that we will not ever touch pool buffers. Does it make sense and is that more secure? I think most of the code is not using "big" arrays (say with length more than 128), and we will hit stackalloc in most cases.
> 
> ```
> Span<byte> arr = length <= 128
>     ? stackalloc byte[128]
>     : new byte[length];
> ```

DeagleGross committed:
dont rent (only stackalloc or allocate new byte array) + address other comments + distinguish .net10 and .netFx impls

DeagleGross commented:
Here is the progress:
1) changed the code to not rent buffers, but either `stackalloc` or allocate a new array
2) changed to proposed stackalloc pattern by Levi
3) distinguished net10 and netFx flows for clarity
4) created an API proposal for SymmetricAlgorithm using `Span<byte>` as a key: https://github.com/dotnet/runtime/issues/111154
5) found out that validation algorithms can be improved to use `Span<byte>` as well if we use a specific implementation reference (like HMACSHA512). If we support enough of known types, we would cover most of the cases IMO.

@GrabYourPitchforks let me know what you think about №1, 4, 5 please
cc @mgravell

vcsjones commented on src/DataProtection/DataProtection/src/SP800_108/ManagedSP800_108_CTR_HMACSHA512.cs:
```diff
@@ -55,6 +57,217 @@ public static void DeriveKeys(byte[] kdk, ArraySegment<byte> label, ArraySegment
         }
     }
 
+#if NET10_0_OR_GREATER
```
.NET already has SP800108 in the box. We should probably just use the one that is built-in to .NET if it is available.

https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.sp800108hmaccounterkdf

DeagleGross replied:
I dont think the API is having the same input \ output args, and I cant understand if I can use the in-box version for the code purposes: destination is being broken apart into `operationSubKey` and `validationSubKey` - could I just slice the `destination` span to get the same data?

sorry, not knowing much about cryptography so can be asking something weird :)

BrennanConroy replied:
Ooh that looks great. Since it has netstandard and netfx targets we could remove a bunch of code in this repo. Should figure out the `operationSubKey`/`validationSubKey` question you have.

DeagleGross replied:
I have tested against runtime provided SP800108, and perf is slower. I will probably raise an issue for that in dotnet/runtime in some time.

For now keeping this usage

danmoseley commented:
> Re №1: Could you please describe how attack surface of the application is increased if pool buffers are used? Does that mean that pooling is easier to inject into via reflection for example?

I'm interested in the answer to this, my assumption was it's just more copies in the heap, so higher probability of appearing in a dump of the service

GrabYourPitchforks commented:
> Could you please describe how attack surface of the application is increased if pool buffers are used?

Using pooled arrays increases the attack surface by allowing other components of the system to have managed references to sensitive key material. It is unfortunately quite common to see buggy components misuse the array pool. And to be fair, the bug would be in _those_ components, not within ASP.NET. And the consequences of such bugs span the gamut of issues - data corruption, information disclosure, etc.

If we allow the pool to contain sensitive data, then the stakes are raised, as it were. These buggy components might now inadvertently disclose sensitive plaintext data - or even _active key material_. Leaking the latter especially would cause the collapse of other security features within ASP.NET applications, including allowing attackers to carry our CSRF attacks or even mint their own administrative auth tokens.

Pinning is similar. By keeping the sensitive data contained within a pinned array, we ensure that no copies of it are made before we have a chance to clear it, which limits the chances of buggy code (either through direct memory access or by leaking uninitialized memory from the array pool) to disclose it.

To be sure, this is all defense-in-depth. In an otherwise properly functioning application, the security of DataProtection does not rely on any of this. These behaviors are solely to minimize the chances of _other_ bugs in the system from manifesting in such a way that it undermines DataProtection. If you'd like to think of it another way: the information processed by DataProtection is so sensitive and failure would be so wildly catastrophic that this is an attempt to isolate the component from other failures within the same process.

Any perf-vs-security tradeoff is always at the discretion of the feature team. And the metrics used to judge these tradeoffs are obviously subjective! But it is nevertheless a tradeoff, and the product team needs to stand behind a statement akin to:

> We do not believe that these protections provide significant enough value to our customers when weighed against the _xx_% performance gain (or other tangible benefit) that foregoing these protections would offer.

--

Fun fact! The original design of this component didn't even store the keys in-proc. It actually stored the keys out-of-proc to further reduce the risk of a buggy component within the process leaking keys. The only reason we didn't move forward with it at GA is that it didn't work in Azure Web Sites. At the time, their sandbox saw our IPC mechanism as a potential threat, and it protected the stability of the machine by throttling our IPC channel to 100 transactions per second. That's 100 `Protect` or `Unprotect` operations per second -- max.

The design was _very_ secure, but it was unworkably slow to the point where no real-world customer could ever depend on it. So for the sake of practicality we ended up with the design we have today.

DeagleGross committed:
Merge branch 'main' into dmkorolev/dataprotection/lin-perf-2

Rick-Anderson commented on src/DataProtection/Cryptography.Internal/src/CryptoUtil.cs:
```diff
@@ -91,6 +91,21 @@ public static bool TimeConstantBuffersAreEqual(byte* bufA, byte* bufB, uint coun
+    [MethodImpl(MethodImplOptions.NoInlining | MethodImplOptions.NoOptimization)]
+    public static bool TimeConstantBuffersAreEqual(ReadOnlySpan<byte> bufA, ReadOnlySpan<byte> bufB)
+    {
+        // Technically this is an early exit scenario, but it means that the caller did something bizarre.
+        // An error at the call site isn't usable for timing attacks.
```
```suggestion
This early exit handles unexpected input without introducing timing vulnerabilities.
```

DeagleGross replied:
can you please elaborate? I should not do `Assert()` here?

Rick-Anderson replied:
> can you please elaborate? I should not do `Assert()` here?


Are you asking me? 
I have no objection to using `Assert`

I replaced your colloquial comments with  formalized/professional comments. 
If my changes are inaccurate, modify my suggestion. 

`Technically` adds no value. The MS style guide mantra is less is more. Bizarre is colloquial and it's ***NOT*** a verb used in computer science to describe an invalid state.

Rick-Anderson replied:
@BrennanConroy  can you suggest better comments?  I might be misunderstanding @DeagleGross  code comments.

DeagleGross replied:
I thought you are proposing something about the code itself.
Comments were just copied from the other method (.net fx implementation).

i will adjust them, thanks for bringing this in

DeagleGross replied:
changed the wording

Rick-Anderson reviewed (changes requested):
Less is more.

BrennanConroy reviewed (commented):
I'm not a fan of duplicating the major crypto methods, leads to a higher maintenance cost. Hopefully we only have a couple differences in each method between TFMs and we can ifdef just those specific sections.

DeagleGross commented on src/DataProtection/DataProtection/src/Managed/ManagedGenRandomImpl.cs:
```diff
@@ -16,6 +17,10 @@ private ManagedGenRandomImpl()
     }
 
+#if NET10_0_OR_GREATER
+    public void GenRandom(Span<byte> target) => RandomNumberGenerator.Fill(target);
```
other `GenRandom` is used in various other places:
![image](https://github.com/user-attachments/assets/d267986a-9d2f-4549-97e5-d7e21d3c35d2)

I dont want to touch other code changing how the allocations happen there and etc. 
Moreover, cant understand how I can terminate other method, because .netfx does not have `Span<>` type. 

Can you please elaborate?

BrennanConroy replied:
Those call sites would just change to something like
```charp
var keyModifier = new byte[KEY_MODIFIER_SIZE_IN_BYTES];
_genRandom.GenRandom(keyModifier.AsSpan());
```

DeagleGross replied:
I just realized - it has the same story. `GenRandom` with `Span` arg implementation would use `RandomNumberGenerator.Fill(Span<byte> bytes)` overload, which is not available in the netFx:
![image](https://github.com/user-attachments/assets/c5ba4cfc-81b7-47c7-8922-d0bc46a9d6c1)

resolving this comment as well. Reopen if you have other propositions!

DeagleGross committed:
address PR comment

DeagleGross commented on src/DataProtection/Cryptography.Internal/src/CryptoUtil.cs:
```diff
@@ -91,6 +91,21 @@ public static bool TimeConstantBuffersAreEqual(byte* bufA, byte* bufB, uint coun
 #endif
     }
 
+#if NET10_0_OR_GREATER
```
there is no `ReadOnlySpan<>` in netFx (https://referencesource.microsoft.com/#q=readonlyspan). So i cant just remove this overload. However, changed to invoke this code from other `TimeConstantBuffersAreEqual` overload to avoid some duplication

DeagleGross replied:
oh, i forgot net api catalog exists - very easy to check if types exist in specific framework and how to get them, thanks!
still [RandomNumberGenerator](https://referencesource.microsoft.com/#mscorlib/system/security/cryptography/randomnumbergenerator.cs,828597225647bdbb) does not have the API for `Span<>` in netFX. so even if I reference Span, then I still cant use the desired API. 

Is there a workaround for this one?

BrennanConroy replied:
😢 Guess not for GenRandom

DeagleGross replied:
resolving a comment then, let me know if you have other options here

DeagleGross commented:
/azp run

DeagleGross committed:
use ReadOnlySpan only in CryptoUtils

DeagleGross committed:
move to "manual" section

DeagleGross committed:
reduce to single implementation ManagedSP800_108_CTR_HMACSHA512.DeriveKeys

DeagleGross committed:
adjust a comment

DeagleGross committed:
change decrypt to single implementation

DeagleGross committed:
single impl for encrypt()

DeagleGross committed:
no pool at all

DeagleGross commented on src/DataProtection/DataProtection/src/Managed/ManagedAuthenticatedEncryptor.cs:
```diff
@@ -186,82 +193,144 @@ public byte[] Decrypt(ArraySegment<byte> protectedPayload, ArraySegment<byte> ad
-            ArraySegment<byte> keyModifier = new ArraySegment<byte>(protectedPayload.Array!, keyModifierOffset, ivOffset - keyModifierOffset);
-            var iv = new byte[_symmetricAlgorithmBlockSizeInBytes];
-            Buffer.BlockCopy(protectedPayload.Array!, ivOffset, iv, 0, iv.Length);
+            ReadOnlySpan<byte> keyModifier = protectedPayload.Array!.AsSpan(keyModifierOffset, ivOffset - keyModifierOffset);
```
i have refactored code - there is a single implementation per method and the variety comes from #if def statements. Please take a look - i hope it much more clear now.

DeagleGross committed:
remove leftover

DeagleGross committed:
oops stackoverflow

DeagleGross committed:
merge main

DeagleGross committed:
correct merge

BrennanConroy commented on src/DataProtection/DataProtection/src/SP800_108/ManagedSP800_108_CTR_HMACSHA512.cs:
```diff
@@ -41,25 +94,64 @@ public static void DeriveKeys(byte[] kdk, ArraySegment<byte> label, ArraySegment
+#if NET10_0_OR_GREATER
+            if (prfOutput is not null)
+            {
+                Array.Clear(prfOutput, 0, prfOutput.Length); // contains key material, so delete it
```
This used to be in the loop, any reason it was moved out? It's not deleting all key material now.

DeagleGross replied:
for netFx or netstandard reverted to clear every iteration (because we allocate every time). for .net10 I am doing clear in the end, because i am reusing the same underlying array for each loop iteration.

DeagleGross committed:
merge main

DeagleGross committed:
no-alloc setKey for algorithm

DeagleGross committed:
raise stackalloc to 256

DeagleGross committed:
Revert "raise stackalloc to 256"

This reverts commit d87aeafebe7239d920173a30a9b408ee56b275aa.

DeagleGross committed:
address ManagedSP800_108_CTR_HMACSHA512

DeagleGross commented on src/DataProtection/DataProtection/src/SP800_108/ManagedSP800_108_CTR_HMACSHA512.cs:
```diff
@@ -41,25 +94,64 @@ public static void DeriveKeys(byte[] kdk, ArraySegment<byte> label, ArraySegment
+#if NET10_0_OR_GREATER
+                // not using stackalloc here, because we are in a loop
+                // and potentially can exhaust the stack memory: https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2014
+                prfOutput = new byte[prfOutputSizeInBytes];
```
for net10 moved to before the loop.

adityamandaleeka commented:
Just saw the latest update: 

> Which provide an evidence of ~10% of max app allocation size and ~46% RPS improvement.

:eyes: nice work!

DeagleGross committed:
Merge branch 'main' into dmkorolev/dataprotection/lin-perf-2

BrennanConroy commented on src/DataProtection/DataProtection/src/Managed/ManagedAuthenticatedEncryptor.cs:
```diff
@@ -272,62 +295,143 @@ public byte[] Decrypt(ArraySegment<byte> protectedPayload, ArraySegment<byte> ad
     {
         plaintext.Validate();
         additionalAuthenticatedData.Validate();
+        var plaintTextSpan = plaintext.AsSpan();
```
```suggestion
        var plaintextSpan = plaintext.AsSpan();
```

BrennanConroy commented on src/DataProtection/DataProtection/src/Managed/ManagedAuthenticatedEncryptor.cs:
```diff
@@ -272,62 +295,143 @@ public byte[] Decrypt(ArraySegment<byte> protectedPayload, ArraySegment<byte> ad
+                    // Step 2: Copy the key modifier and the IV to the output stream since they'll act as a header.
+                    keyModifier.CopyTo(outputSpan.Slice(start: 0, length: keyModifierLength));
+
+                    // Step 3: Generate IV for this operation right into the result array (no allocation)
```
nit: step 3 is part of step 2, reword or remove

BrennanConroy commented on src/DataProtection/DataProtection/src/Managed/ManagedAuthenticatedEncryptor.cs:
```diff
@@ -272,62 +295,143 @@ public byte[] Decrypt(ArraySegment<byte> protectedPayload, ArraySegment<byte> ad
+                    keyModifier.CopyTo(outputSpan.Slice(start: 0, length: keyModifierLength));
+
+                    // Step 3: Generate IV for this operation right into the result array (no allocation)
+                    _genRandom.GenRandom(outputSpan.Slice(start: keyModifierLength, length: ivLength));
+                    var iv = outputSpan.Slice(start: keyModifierLength, length: ivLength);
```
```suggestion
                    var iv = outputSpan.Slice(start: keyModifierLength, length: ivLength);
                    _genRandom.GenRandom(iv);
```

DeagleGross committed:
merga main

DeagleGross committed:
address PR comments

DeagleGross committed:
merge main

BrennanConroy commented on src/DataProtection/DataProtection/src/SP800_108/ManagedSP800_108_CTR_HMACSHA512.cs:
```diff
@@ -41,25 +100,48 @@ public static void DeriveKeys(byte[] kdk, ArraySegment<byte> label, ArraySegment
+        {
+            if (prf is IDisposable disposablePrf)
+            {
+                disposablePrf.Dispose();
```
I think this should be removed now

BrennanConroy commented on src/DataProtection/DataProtection/test/Microsoft.AspNetCore.DataProtection.Tests/E2ETests.cs:
```diff
@@ -0,0 +1,28 @@
+public class E2ETests
+{
+    [Fact]
+    public void ProtectAndUnprotect_ForSampleAntiforgeryToken()
```
But you're just calling protect then unprotect on the value, so all you're really testing is a random string is round trip-able which I'm sure is already tested.

BrennanConroy commented on src/DataProtection/DataProtection/src/Managed/ManagedAuthenticatedEncryptor.cs:
```diff
@@ -144,20 +147,28 @@ private SymmetricAlgorithm CreateSymmetricAlgorithm()
     }
 
-    private KeyedHashAlgorithm CreateValidationAlgorithm(byte[] key)
+    private KeyedHashAlgorithm CreateValidationAlgorithm()
```
Nit: Remove? The other method uses a default value so should work for all current callsites.

DeagleGross replied:
ye, removed

DeagleGross committed:
remove test + address nits

DeagleGross commented:
just as a history note: @BrennanConroy noticed that 111k RPS was before the openssl update, so we dont see 45% RPS improvement, but >10%
![image](https://github.com/user-attachments/assets/2bf6b7d2-db90-4011-827c-08056a92d1dd)

however allocation rate is still going lower significantly even for a 15 sec run:
![image](https://github.com/user-attachments/assets/11722df0-c46e-42ff-bd63-92d0679ddfce)

cc @adityamandaleeka
